src/t_stream.c

```diff
@@ -1350,6 +1367,155 @@ robj *createObjectFromStreamID(streamID *id) {
                         id->ms,id->seq));
 }
 
+/* Returns non-zero if the ID is 0-0. */
+int streamIDEqZero(streamID *id) {
+    return !(id->ms || id->seq);
+}
+
+/* A helper that returns non-zero if the range from 'start' to the stream's
+ * end doesn't contain a tombstone.
+ *
+ * NOTE: this assumes that the caller had verified that 'start' is less than
+ * 's->last_id'. */
+int streamIsContiguousRange(stream *s, streamID *start) {
+    streamID start_id;
+
+    if (!s->length || streamIDEqZero(&s->xdel_max_id)) {
+        /* An empty stream or no tombstones. */
+        return 1;
+    }
+
+    /* Copy start ID, if given, or default to 0-0. */
+    if (start) {
+        start_id = *start;
+    } else {
+        start_id.ms = 0;
+        start_id.seq = 0;
+    }
+
+    if (streamCompareID(&s->first_id,&s->xdel_max_id) > 0) {
+        /* The latest tombstone is before the first entry. */
+        return 1;
+    }
+
+    if (streamCompareID(&start_id,&s->xdel_max_id) >= 0) {
+        /* The range doesn't include a tombstone. */
+        return 1;
+    }
+
+    /* The range includes a tombstone. */
+    return 0;
+}
+
+/* Replies with a consumerg group's current lag, that is the number of messages
+ * in the stream that are yet to be delivered. In case that the lag isn't
+ * available due to fragmentation, the reply to the client is a null. */
+void streamReplyWithCGLag(client *c, stream *s, streamCG *cg) {
+    int valid = 1;
+    uint64_t lag = 0;
+
+    if (!s->offset) {
+        /* The lag of a newly-initialized stream is 0. */
+        lag = 0;
+    } else if (cg->offset && streamIsContiguousRange(s,&cg->last_id)) {
+        /* No fragmentation ahead means that the group's offset is valid for
+         * performing the lag calculation. */
+        lag = s->offset - cg->offset;
+    } else if (streamIDEqZero(&cg->last_id)) {
+        if (streamIsContiguousRange(s,NULL)) {
```

Comment:
can you please add `serverAssert(cg->offset == 0);` ?

Reply:
@guybe7 `XGROUP SETID mystream mygroup 0 1` will cause `last_id` to be zero, and `offset` is 1.
@itamarhaber I'm confused about the `offset` param in `xgroup setid`, how do users know what the `offset` of `last_id`.

Reply:
Resolved.

@sundb they don't have to know - they can put any positive integer there, and the logic sanitizes it (i.e. it must be the before first, first or last IDs).